vm/storage.go

```diff
@@ -109,8 +111,14 @@ func (s Storage) Put(v interface{}) (cid.Cid, error) {
 	if blk, ok := v.(blocks.Block); ok {
 		// optimize putting blocks
 		nd, err = cbor.DecodeBlock(blk)
-	} else if bytes, ok := v.([]byte); ok {
-		nd, err = cbor.Decode(bytes, types.DefaultHashFunction, -1)
+	} else if raw, ok := v.([]byte); ok {
+		nd, err = cbor.Decode(raw, types.DefaultHashFunction, -1)
+	} else if cm, ok := v.(cbg.CBORMarshaler); ok {
```

Comment:
Are we putting `CBORMarshaler`s into storage? Is this new functionality? If so, and we aren't yet using it, we should wait to introduce this until it is exercised by code (and tests).

Reply:
so a CBORMarshaler just means an object that conforms to that interface, meaning it has methods to fast path it's own serialization / deserialization -- this covers all hamt nodes, and in fact, they need to be fastpathed to encode properly. WrapObject won't work here because of the cbg.Deferred object. The actual solution is that go-ipld-cbor should check for fast path in its own calls to marshal/unmarshal, which there is an unmerged PR for: https://github.com/ipfs/go-ipld-cbor/pull/64 . Absent that getting merged, this is neccesary for a couple direct calls to storage.Put

Reply:
^^ @acruikshank

Reply:
will add a comment to that effect.